### PR TITLE
mon/OSDMonitor: kill pending upmap changes too if pool is gone

### PR DESCRIPTION
**Pull request by xiexingguo**
There are quite a lot mgr/balancer related tests that I can observe
the following logged errors:
```
2018-03-03 08:14:00.359946 7fdda18cc700 -1 maybe_remove_pg_upmaps unable to load crush-rule of pg 5.b
```
which turns out to be a *pool-deletion* vs *balancer-auto-injected-upmap-changes*
race issue.
The root cause is that we don't clean up those *pending*
pg_upmap/pg_upmap_items injected by the mgr/balancer properly simultaneously
when the corresponding pool is gone, and hence the above problem can be fixed by:
1. clean up any pending upmap changes too if the corresponding pool is gone
2. re-check pending pool removal queue before we can safely apply any new upmap changes

Signed-off-by: xie xingguo <xie.xingguo@zte.com.cn>

### CODE REVIEW

**Comment by xiexingguo** [comment]
@liewegas  Can you review this too? Thanks!

File: src/mon/OSDMonitor.cc [hunk 1]
```diff
@@ -9540,6 +9540,11 @@ bool OSDMonitor::prepare_command_impl(MonOpRequestRef op,
       err = -ENOENT;
       goto reply;
     }
+    if (pending_inc.old_pools.count(pgid.pool())) {
+      ss << "pool of " << pgid << " is pending removal";
+      err = -ENOENT;
+      goto reply;
```
Comment (liewegas) [inline src/mon/OSDMonitor.cc]: instead of goto reply this should wait for the commit and then return the error (the pool deletion hasn't committed yet)

Comment (xiexingguo) [reply src/mon/OSDMonitor.cc]: fixed

**Comment by tchaikov** [comment]
retest this please

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-03-08_12:59:37-rados-wip-kefu-testing-2018-03-08-1932-distro-basic-smithi/

- the swift failure was caused by #20419
- and i reran `osd-pool-create.sh` multiple times locally, and was not able to reproduce that failure.

**Pull request merged**